core/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java

```diff
@@ -131,6 +153,15 @@ public WritePrimaryResult(Request request, Response finalResponse,
             postWriteActions(indexShard, request, location, this, logger);
         }
 
+        public WritePrimaryResult(Request request, Exception finalFailure, IndexShard indexShard) {
+            super(finalFailure);
+            /*
+             * We call this before replication because this might wait for a refresh and that can take a while. This way we wait for the
+             * refresh in parallel on the primary and on the replica.
+             */
+            postWriteActions(indexShard, request, null, this, logger);
```

Comment:
why do we need this when failing and not replicating?

Also - unrelated but will make this easier to understand. Can we make TransportWriteAction use the same distinction between primary and replica request on it's types?